Blog for search update (#73)
readthedocs/blog: sphinx-search-as-you-type.rst

.. post:: July 15, 2019
   :tags: gsoc, search, feature, in-doc-search
   :author: Vaibhav
   :location: LKO

> Search Update

Reviewer: This could definitely use a better title. We should mention what we updated, instead of being vague.

Author: I am not sure about what would be a better title.
What about: `Search Update: Section Linking & Search As You Type Feature`

Reviewer 2: I think search as you type is the biggest feature and it should definitely be in the title. I wouldn't be opposed to "Adding Search as You Type to Read the Docs" or something like that.

Reviewer: Yea, I'd like to make this title better before merging.

Author: I don't have very strong opinion but I do think that `Adding Search as You Type to Read the Docs` focuses only on **one** feature?

Reviewer: "Search Update" means nothing, we need to tell people what was updated. I don't feel strongly what we focus on, but it needs more.

Author: I am not in favor of `Search Update` also.
How about -- `Adding Search as You Type to Read the Docs and much more`?
Or we can just go with `Adding Search as You Type to Read the Docs`.

Reviewer: `GSOC 2019: Improved Search Results and Search As You Type`?

Author: Loved that one. :heart:

Author: Updated

=============

Giving users the ability to easily find the information that they
are looking for has always been important for Read the Docs.
This year, I, `Vaibhav Gupta`_, took the opportunity provided
by Google Summer of Code to improve the search.
The main goal of my `GSoC project`_ was to enable multiple hits per search result,
improve the UI/UX around code search and to make an `Sphinx extension`_ to provide
"search as you type" experience to the users.

> 

Reviewer: We should add another short paragraph here mentioning that it's live on our docs, and including a link so users can go try it out themselves.

---------------------

Google Summer of Code is a global program where students work with
an open source organization on a 3-month programming project.
I got accepted in the program and was excited to spend my summer with Read the Docs.

I have worked full time during the GSoC coding period and implemented various features.
All of my search related work can be seen in the `In-Doc Search UI Project Board`_.

Background
----------

Search code was voluntarily contributed by `Rob Hudson`_,
`back in 2013`_ and then improved by other contributors.
It was greatly improved and upgraded by `Safwan Rahman`_ during :doc:`GSoC'18 <search-improvements>`.
Continuining on the same path,
I have implemented some new features on top of the existing search backend.

New Features
------------

During the GSoC period, I have worked on the following features:

- `In-Doc Search UI`_: We have worked on a Sphinx extension which adds a clean and minimal
  search UI to your docs. It supports "search as you type" feature.
  So, it is now possible to get instant results without being redirected to any other page.
  Read the docs `here`_.

.. image:: /_static/in-doc-search-demo.gif
    :width: 100%

>     :target: /_static/in-doc-search-demo.gif

Reviewer: I love the demo, but I think we should also link to some example searches in our docs for each of these features that show it off.

- Multiple Hits Per Search Result: This is one of the highly requested features.
  Now, we support results from the sections of the docs, clicking on which will take you
  to that particular section and not just to the top of the result page.

>   You don't have to scroll anymore to the desired result.

Reviewer: We can remove this line I think, it's mostly just stating what we already said in another way

- Code Search: We now support code search. If you want to search a particular function
  or an API endpoint -- you can just type your query and you will find it in the results.
  Eg: ``api/v3/`` or ``module.function``.

What Next?
----------

We don't intend to stop just yet.
We are planning to work on some more cool features in the near future,
some of which are:

- **Search Facets**:
  Facets can be used to make search more accurate.

>   Like ``subproject: project query``.

Reviewer: This could use a real example. For example on the Celery docs you could search inside django-celery for `settings` or something.

Author: I wasn't able to find any search results from django-celery in the celery docs, but i do find results from Kombu, so I included its example.

- **Search Analytics**:
  Analytics is always helpful to know what your users are actually looking for in your docs.
- **Search Results Ordered By Most Viewed Pages**:
  It would be much more useful if the most viewed pages are shown first in the search results.
- **Search Inside Sections**:
  It would be good if users have the option to get the
  search results from a particular section of the documentation.

>   (`readthedocs/readthedocs-sphinx-search#23`_).

Reviewer: It's great this has an issue attached. We should probably create issues for the other ideas mentioned in this post so people can follow along if they are interested. It would also help with contribution.

Contributors Wanted
-------------------

As Read the Docs is an open source project backed by a small team of developers,

> most of them are busy to keep things up and running only.

Reviewer: ```suggestion
most of them are busy just keeping the site up and running.
```

Therefore, it's quite hard for them to take time to implement new features.
If you know some bit of Django or Python and Elasticsearch,
you can contribute to the search functionality of Read the Docs.
If you need any support to start contributing,
you can get in touch with me or any member of Read the Docs team.
You can find all of us at *#readthedocs* freenode IRC channel or `readthedocs gitter`_ channel.
I am *dojutsu-user* at IRC and *@dojutsu-user* at gitter.

Conclusion
----------

> To conclude, I must say, these new features will make it

Reviewer: ```suggestion
These new features will make it much 
```

easier to find the relevant information in the docs.
There are an infinite number of ways it can be improved and I believe we can compete
with major search engines in terms of documentation searching.
We don’t need superhero or coding guru, just need people who understand Python,
Django and Elasticsearch and have some time to write some code for us.

> You are a **Superhero** to us if you can lend your time and effort to improve Read the Docs.

Reviewer: ❤️

.. post:: August 09, 2019
   :tags: gsoc, search, feature, in-doc-search, analytics
   :author: Vaibhav
   :location: LKO

Search Update
=============

Giving users the ability to easily find the information that they
are looking for has always been important for Read the Docs.
This year, I, `Vaibhav Gupta`_, took the opportunity provided
by Google Summer of Code to improve the search.
The main goals of my `GSoC project`_ were:

- to make an `Sphinx extension`_ to provide "search you type" experience to the users.
- to add support for multiple hits per search result.
- to improve the UI/UX around code search.
- to add support for search analytics.

Search as you type feature is live on our docs.

> You can go through these links to experience it.

Reviewer: ```suggestion
You can go through these links to experience it:
```

- https://docs.readthedocs.io/?rtd_search=contributing
- https://docs.readthedocs.io/?rtd_search=api/v3/projects/

Google Summer of Code
---------------------

Google Summer of Code is a global program where students work with
an open source organization on a 3-month programming project.
I got accepted in the program and was excited to spend my summer with Read the Docs.

I have worked full time during the GSoC coding period and implemented various features.
All of my search related work can be seen in the `In-Doc Search UI Project Board`_.

Background
----------

> Search code was voluntarily contributed by `Rob Hudson`_,

Reviewer: ```suggestion
The search code was originally contributed by `Rob Hudson`_,
```

  So, it is now possible to get instant results without being redirected to any other page.
  Read the docs `here`_.

.. image:: /_static/in-doc-search-demo.gif
    :width: 100%
    :target: /_static/in-doc-search-demo.gif

- **Multiple Hits Per Search Result**: This is one of the highly requested features.
  We now support search results from the sections of the docs, clicking on which will take you
  to that particular section and not just to the top of the result page.

- **Code Search**: We now support code search. If you want to search a particular function
  or an API endpoint -- you can just type your query and you will find it in the results.
  Eg: ``api/v3/`` or ``module.function``.

- **Search Analytics**: We now have support for search analytics.
  These analytics makes it easy to know what the users are looking for in your documentation.
  You can see these analytics in your project admin dashboard.
  Currently, this feature is in beta state and and is available under a `feature flag`_.
  We plan to make this available for everyone soon.
  If you want to test this feature out and help giving us feedback,

>   please contact us via `GitHub issues`_.

Reviewer: This should have a screenshot, also.

  Currently, this feature is in beta state and is available under a `feature flag`_.
  We plan to make this available for everyone soon.
  If you want to test this feature out and help giving us feedback,
  please contact us via `GitHub issues`_.

.. figure:: /_static/search-analytics-demo.png
    :width: 60%
    :align: center
    :target: /_static/search-analytics-demo.png
    :alt: Search analytics in project admin dashboard

    Search analytics demo dashboard

What Next?
----------

We don't intend to stop just yet.
We are planning to work on some more cool features in the near future,
some of which are:

- **Search Facets**:
  Facets can be used to make search more accurate.
  For example: In `Celery docs`_, facets can be used to search inside `Kombu docs`_ for "serializers",
  like ``subproject: kombu serializers``.
  (`readthedocs/readthedocs.org#5966`_)
- **Search Results Ordered By Most Viewed Pages**:
  It would be much more useful if the most viewed pages are shown first in the search results.
  (`readthedocs/readthedocs.org#5968`_)
- **Search Inside Sections**:
  It would be good if users have the option to get the

>   search results from a particular section of the documentation.

Reviewer 2: From this description, this isn't totally clear what it would do.

Author: I added an example.
